File: meldingsvarsler-fasit-forslag/src/main/java/no/sonat/meldingsvarsler/MeldingSender.java
package no.sonat.meldingsvarsler;

import no.sonat.meldingsvarsler.infrastructure.abonnent.AbonnentRepositoryReader;
import no.sonat.meldingsvarsler.infrastructure.meldinger.MeldingRepository;
import no.sonat.meldingsvarsler.meldinger.MeldingProsessor;

import java.util.List;

public class MeldingSender {
    private final MeldingRepository meldingRepository;
    private final AbonnentRepositoryReader abonnentRepository;
    private final List<MeldingProsessor> meldingProsessorer;

    /*
    Dependency Inversion Principle :
    abonnentRepository, meldingRepository, meldingProsessors
    */
    public MeldingSender(MeldingRepository meldingRepository,
            AbonnentRepositoryReader abonnentRepository,
            List<MeldingProsessor> meldingProsessors) {
        this.abonnentRepository = abonnentRepository;
        this.meldingProsessorer = meldingProsessors;
        this.meldingRepository = meldingRepository;
    }

    public void sendMeldinger() {
        meldingRepository.hentMeldinger().forEach(melding -> {
            meldingProsessorer.stream()
                .filter(meldingProsessor -> meldingProsessor.haandtererUtsendingAv(melding))
                .forEach(meldingProsessor -> {
                    abonnentRepository.hentAbonnenter().stream()
                            .filter(abonnent -> abonnent.abonnererPaa(melding))
                            .forEach(abonnent -> meldingProsessor.sendMelding(abonnent, melding));
                });
        });
    }
}

File: meldingsvarsler-fasit-forslag/App/MeldingSender.cs
using System.Collections.Generic;
using System.Linq;

namespace App {
    public class MeldingSender
    {
        readonly IMeldingRepository meldingRepository;
        readonly IAbonnentRepositoryReader abonnentRepository;
        readonly List<IMeldingProsessor> meldingProsessorer;

        /* Inversion Of Control : abonnentRespository, meldingRepository, meldingProsessors */
        public MeldingSender(IMeldingRepository meldingRepository,
                             IAbonnentRepositoryReader abonnentRepository,
                             List<IMeldingProsessor> meldingProsessorer)
        {
            this.abonnentRepository = abonnentRepository;
            this.meldingProsessorer = meldingProsessorer;
            this.meldingRepository = meldingRepository;
        }

        public void SendMeldinger()
        {
            meldingRepository.HentMeldinger().ForEach(melding =>
            {
                meldingProsessorer.Where(meldingProsessor => meldingProsessor.HaandtererUtsendingAv(melding))
                                  .ToList()
                                  .ForEach(meldingProsessor =>
                                  {
                                      abonnentRepository.HentAbonnenter()
                                                        .Where(abonnent => abonnent.AbonnererPaa(melding))
                                                        .ToList()
                                                        .ForEach(abonnent =>
                                                                     meldingProsessor.SendMelding(abonnent, melding));
                                  });
            });
        }
    }
}